components/DataLiberation/URL/class-cssurlprocessor.php

```diff
@@ -0,0 +1,464 @@
+<?php
+
+namespace WordPress\DataLiberation\URL;
+
+use WP_HTML_Text_Replacement;
+
+use function WordPress\Encoding\codepoint_to_utf8_bytes;
+
+/**
+ * Finds and replaces URLs declared using a url() notation
+ * in a CSS block body (without the trailing braces). An
+ * example of such a block body is the content of a style=""
+ * HTML attribute.
+ *
+ * This class was initially created to migrate background-image
+ * URLs in CSS blocks during a WXR import.
+ *
+ * Usage:
+ *
+ * ```php
+ * $css_block_body = <<<CSS
+ * /* John picked this photo: *\/
+ * background: url("/pictures/johns-pic.jpg");
+ * content: "Ever heard about url() notation? Like this: url(/jane/picture.jpg)";
+ * CSS;
+ *
+ * $processor = new CSSUrlProcessor( $css_block_body );
+ * while ( $processor->next_url() ) {
+ *     $processor->set_raw_url( '/new-image.jpg' );
+ * }
+ * echo $processor->get_updated_css();
+ * ```
+ *
+ * Output:
+ *
+ * ```php
+ * /* John picked this photo: *\/
+ * background: url("/new-image.jpg");
+ * content: "Ever heard about url() notation? Like this: url(/jane/picture.jpg)";
+ * ```
+ */
+class CSSUrlProcessor {
+
+	/**
+	 * The CSS block to process (without the trailing braces).
+	 *
+	 * @var string
+	 */
+	private $css;
+	private $url_starts_at;
+	private $url_length;
+	private $bytes_already_parsed = 0;
+
+	/**
+	 * @var string
+	 */
+	private $matched_url;
+
+	/**
+	 * @var string
+	 */
+	private $decoded_url;
+
+	/**
+	 * @see \WP_HTML_Tag_Processor
+	 * @var WP_HTML_Text_Replacement[]
+	 */
+	private $lexical_updates = array();
+
+	public function __construct( $css ) {
+		$this->css = $css;
+	}
+
+	/**
+	 * Finds the next URL in the CSS content.
+	 *
+	 * Uses a state machine parser to handle arbitrarily large data URIs (1MB+)
+	 * which would otherwise hit PCRE limits.
+	 *
+	 * @return bool True if a URL was found, false otherwise.
+	 */
+	public function next_url() {
+		$this->matched_url   = null;
+		$this->decoded_url   = null;
+		$this->url_starts_at = null;
+		$this->url_length    = null;
+
+		$length = strlen( $this->css );
+		$at     = $this->bytes_already_parsed;
+
+		while ( $at < $length ) {
+			// Optimization: Use strcspn to skip to next interesting character in one pass.
+			// Look for: u (start of url), / (comment), " (string), ' (string).
```

Comment:
might this lead to dispersing when the `style` value is broken? I believe that there are specific error-handling behaviors, but I don’t know if skipping to these characters would bypass that. I hope this is safe because it would likely mean a significant slow-down to go token-by-token.